File: ts/earthbucks-lib/src/iso-hex.ts
import { EbxBuf } from "./ebx-buf";
import { Result, Ok, Err } from "earthbucks-opt-res";
import { EbxError, InvalidHexError } from "./ebx-error.js";
import { Option, Some, None } from "earthbucks-opt-res";

export class IsoHex {
  static isValid(hex: string): boolean {
    return /^[0-9a-f]*$/.test(hex) && hex.length % 2 === 0;
  }

  static encode(buffer: EbxBuf): string {
    return buffer.toString("hex");
  }

  static decode(hex: string): Result<EbxBuf, EbxError> {
    if (!IsoHex.isValid(hex)) {
      return Err(new InvalidHexError(None));
    }
    const buffer = EbxBuf.from(hex, "hex");
    return Ok(buffer);
  }
}

File: rs/earthbucks_lib/src/iso_hex.rs
use hex;
use lazy_static::lazy_static;
use regex::Regex;

use crate::ebx_error::EbxError;

lazy_static! {
    static ref RE: Regex = Regex::new(r"^[0-9a-f]*$").unwrap();
}

pub fn is_valid(hex: &str) -> bool {
    RE.is_match(hex) && hex.len() % 2 == 0
}

pub fn encode(data: &[u8]) -> String {
    hex::encode(data)
}

pub fn decode(hex: &str) -> Result<Vec<u8>, EbxError> {
    if !is_valid(hex) {
        return Err(EbxError::InvalidHexError { source: None });
    }
    let res = hex::decode(hex);
    if res.is_err() {
        return Err(EbxError::InvalidHexError { source: None });
    }
    Ok(res.unwrap())
}
